wcfsetup/install/files/lib/form/MultifactorManageForm.class.php

```diff
@@ -0,0 +1,217 @@
+		}
+		
+		$objectType = ObjectTypeCache::getInstance()->getObjectType(intval($_GET['id']));
+		
+		if (!$objectType) {
+			throw new IllegalLinkException();
+		}
+		if ($objectType->getDefinition()->definitionName !== 'com.woltlab.wcf.multifactor') {
+			throw new IllegalLinkException();
+		}
+		
+		$this->method = $objectType;
+		$this->processor = $this->method->getProcessor();
+		$this->setup = Setup::find($this->method, WCF::getUser());
+		
+		// Backup codes may not be managed if they are not yet set up.
+		if ($this->method->objectType === 'com.woltlab.wcf.multifactor.backup' && !$this->setup) {
+			throw new PermissionDeniedException();
+		}
+	}
+	
+	/**
+	 * @inheritDoc
+	 */
+	protected function createForm() {
+		parent::createForm();
+		
+		$this->processor->createManagementForm($this->form, $this->setup, $this->returnData);
+	}
+	
+	public function save() {
+		AbstractForm::save();
+
+		WCF::getDB()->beginTransaction();
+		
+		/** @var Setup|null $setup */
+		$setup = null;
+		if ($this->setup) {
+			$setup = $this->setup->lock();
+		}
+		else {
+			$setup = Setup::allocateSetUpId($this->method, WCF::getUser());
+		}
+		
+		if (!$setup) {
+			throw new \RuntimeException("Multifactor setup disappeared");
+		}
+		
+		$this->returnData = $this->processor->processManagementForm($this->form, $setup);
+		
+		$this->setup = $setup;
+		WCF::getDB()->commitTransaction();
+		
+		WCF::getDB()->beginTransaction();
+		if (!$this->areBackupCodesSetup()) {
+			$backupMethod = $this->getBackupCodesObjectType();
+			$backupProcessor = $backupMethod->getProcessor();
+
+			// Create Form
+			$form = FormDocument::create('backupCodes');
+			$backupProcessor->createManagementForm($form, null, []);
+			$form->build();
+			
+			// Process Form
+			$form->requestData([
+				'generateCodes' => 'generateCodes',
+			]);
+			$form->readValues();
+			$backupSetupId = Setup::allocateSetUpId($backupMethod, WCF::getUser());
+			$returnData = $backupProcessor->processManagementForm($form, $backupSetupId);
+			$form->cleanup();
+			
+			// Re-create form
+			$form = FormDocument::create('backupCodes');
+			$backupProcessor->createManagementForm($form, $backupSetupId, $returnData);
+			/** @var IFormParentNode $container */
+			$container = $form->getNodeById('existingCodesContainer');
+			$container->insertBefore(
+				TemplateFormNode::create('initialBackup')
+					->templateName('__multifactorManageInitialBackup'),
+				'existingCodes'
+			);
+			$form->build();
+			$this->backupForm = $form;
+		}
+		WCF::getDB()->commitTransaction();
+		
+		$this->saved();
+	}
+	
+	/**
+	 * Returns the Object type representing backup codes.
+	 */
+	protected function getBackupCodesObjectType(): ObjectType {
+		return ObjectTypeCache::getInstance()->getObjectTypeByName('com.woltlab.wcf.multifactor', 'com.woltlab.wcf.multifactor.backup');
+	}
+	
+	/**
+	 * Returns whether backup codes are set up yet.
+	 */
+	protected function areBackupCodesSetup(): bool {
```

Comment:
```suggestion
	protected function hasBackupCodes(): bool {
```
More in line with similar methods and sounds much cleaner.